tests/math_utils.test.ts

```diff
@@ -85,3 +90,128 @@ describe("remap", () => {
     expect(remap(2, 0, 1, 0, 10, false)).to.equal(20);
   });
 });
+
+describe("getFrameSizeInScreenPx", () => {
+  it("returns canvas size when frame is the same aspect ratio", () => {
+    const frameSizePx = getFrameSizeInScreenPx([100, 100], [1, 1], 1);
+    expect(frameSizePx).to.deep.equal([100, 100]);
+  });
+
+  it("does not change returned size if frame resolution is higher", () => {
+    const frameResolutions: [number, number][] = [
+      [0.1, 0.1],
+      [1, 1],
+      [10, 10],
+      [100, 100],
+    ];
+    for (const frameResolution of frameResolutions) {
+      const frameSizePx = getFrameSizeInScreenPx([100, 100], frameResolution, 1);
+      expect(frameSizePx).to.deep.equal([100, 100]);
+    }
+  });
+
+  it("constrains by height when frame aspect ratio is taller than canvas", () => {
+    const canvasSizePx: [number, number] = [100, 100];
+    const frameResolution: [number, number] = [0.5, 1];
+    const frameZoom = 1;
+    const frameSizePx = getFrameSizeInScreenPx(canvasSizePx, frameResolution, frameZoom);
+    expect(frameSizePx).to.deep.equal([50, 100]);
+  });
+
+  it("constrains by width when frame aspect ratio is wider than canvas", () => {
+    const canvasSizePx: [number, number] = [100, 100];
+    const frameResolution: [number, number] = [1, 0.5];
+    const frameZoom = 1;
+    const frameSizePx = getFrameSizeInScreenPx(canvasSizePx, frameResolution, frameZoom);
+    expect(frameSizePx).to.deep.equal([100, 50]);
+  });
+
+  it("scales frame dimensions with zoom", () => {
+    const canvasSizePx: [number, number] = [100, 100];
+    const frameResolution: [number, number] = [1, 0.5];
+    const frameZoom = 2;
+    const frameSizePx = getFrameSizeInScreenPx(canvasSizePx, frameResolution, frameZoom);
+    expect(frameSizePx).to.deep.equal([200, 100]);
+  });
+});
+
+describe("convertCanvasOffsetPxToFrameCoords", () => {
+  const makeConversionTester = (props: {
```

Comment:
should this have an "it" statement around it?

Reply:
ok, I think I'm following what you did here, but it seems like overkill that actually makes the tests harder to read. I really prefer each unit text to test one and only one thing. This seems like you've built a way to test a bunch of different coordinates together?

Reply:
Ah, I made this so I wouldn't have a ton of repeated boilerplate for each test. Each of the unit tests below is checking a single set of parameters (aspect ratio/scale/pan) and checks the four corners and the center of the frame for each.